appcli/variables_manager.py

```diff
@@ -43,86 +55,198 @@ def get_variable(self, path):
         except KeyError as exc:
             raise KeyError(f"Setting [{path}] not set in configuration.") from exc
         return variable
 
     def get_all_variables(self):
-        return self.__get_configuration()
+        return self._get_configuration()
 
     def set_variable(self, path: str, value: Union[str, bool, int, float]):
-        """Sets a value in the configuration.
+        """Sets a value in the configuration
 
         Type of value is not enforced but this might not serialise into yml in a deserialisable format.
 
         Args:
-            path (str): Dot notation for the setting. E.g. insilico.external.database.host
+            path (str): Dot notation for the setting. E.g. settings.insilico.external.database.host
             value: value for the setting
-        """
-        configuration = self.__get_configuration()
-
-        path_elements = path.split(".")
-        parent_path = path_elements[:-1]
-
-        # ensure parent path exists
-        def ensure_path(parent, child):
-            if child not in parent:
-                parent[child] = {}
-            return parent[child]
-
-        reduce(ensure_path, parent_path, configuration)
 
-        # set the value
-        parent_element = reduce(lambda e, k: e[k], parent_path, configuration)
-        parent_element[path_elements[-1]] = value
-
-        self.__save(configuration)
+        TODO: warn user if attempting to save variable outside of the main config file.
+        TODO: remember to do unit testing for nested variables etc...
+        """
+        configuration = self._get_main_configuration()
+        path_elements = path.split(".")  # Divide path into array.
+        # Iterate through each context on path, to create a recursive dictionary.
+        # Main is a pointer to top dictionary object.
+        # Sub is a pointer to the current dictionary layer.
+        dictionary_main = dictionary_sub = {}
+        for field in path_elements:
+            # Apply value to bottom dictionary only.
+            dictionary_sub[field] = value if field is path_elements[-1] else {}
+            dictionary_sub = dictionary_sub[
+                field
+            ]  # Move sub pointer to next layer down.
+
+        def recursive_dictionary_merge(d1: Dict, d2: Dict):
+            # Base case (either d1 or d2 is value).
+            if not hasattr(d1, "copy") or not hasattr(d2, "copy"):
+                return d2  # d2 overwrites fields in d1.
+            d0 = d1.copy()  # New return dictionary.
+            for sub in d2.keys():
+                if sub not in d1.keys():  # Item isn't a duplicate.
+                    d0 |= {sub: d2.get(sub)}
+                else:  # Item is a duplicate.
+                    d0 |= {sub: recursive_dictionary_merge(d1.get(sub), d2.get(sub))}
+            return d0
+
+        self._save(
+            recursive_dictionary_merge(configuration, dictionary_main)[
+                self.configuration_file.stem
```

Comment:
Have investigated - I think it's because of the dictionary nesting. 
`{ 'settings' : { 'a' : 'b' } }` needs to be un-nested to `{ 'a' : 'b' }` for `self._save()` to work.